### PR TITLE
[BugFix][Qwen-Image] align txt_seq_lens RoPE width with padded embeds

### LINKED ISSUE #4443: [Bug] QwenImage RoPE `txt_seq_lens` uses `mask.sum()` instead of padded embed width (misaligned with diffusers)
## Summary

vLLM-Omni QwenImage pipelines set `txt_seq_lens` from `prompt_embeds_mask.sum(dim=1)` when preparing generation context / `prepare_encode()`. Diffusers (latest `main`) instead derives RoPE text length from the **padded encoder hidden-states width** (`encoder_hidden_states.shape[1]`) inside `QwenImageTransformer2DModel.forward()` via `compute_text_seq_len_from_mask()`.

When prompt embeddings are wider than the number of valid (non-padding) tokens — common under **continuous batching (CB)** after in-place padding, RL training with fixed-width collation, or any caller that supplies pre-padded embeds — vLLM-Omni builds a **too-short** text RoPE table. That changes attention numerics relative to diffusers/PyTorch reference and can inflate rollout–trainer KL in RL setups.

Verified on **vllm-omni `main` @ `a693ae67`** against **diffusers `main` @ `d1f8e55c3`**.

## Affected code (vllm-omni)

All QwenImage pipeline variants set `txt_seq_lens` from `mask.sum()` in their generation-context helpers:

- `vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py` (`_prepare_generation_context`, used by `forward()`, `diffuse()`, and `prepare_encode()`)
- `vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image_edit.py`
- `vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image_edit_plus.py`
- `vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image_layered.py`

```python
# Current (vllm-omni):
txt_seq_lens = prompt_embeds_mask.sum(dim=1).tolist() if prompt_embeds_mask is not None else None
negative_txt_seq_lens = (
    negative_prompt_embeds_mask.sum(dim=1).tolist() if negative_prompt_embeds_mask is not None else None
)
```

`prepare_encode()` copies these values onto `DiffusionRequestState.txt_seq_lens` / `negative_txt_seq_lens`. The stepwise CB path (`InputBatch`) reuses them but **does not refresh them** after `_prepare_request_prompt_field()` pads `prompt_embeds` / masks on the request state in place.

## Reference behavior (diffusers latest main)

`diffusers.models.transformers.transformer_qwenimage.compute_text_seq_len_from_mask()` returns the encoder tensor width for RoPE:

```python
batch_size, text_seq_len = encoder_hidden_states.shape[:2]
# ...
return text_seq_len, per_sample_len, encoder_hidden_states_mask
```

`QwenImageTransformer2DModel.forward()` uses that width:

```python
text_seq_len, _, encoder_hidden_states_mask = compute_text_seq_len_from_mask(
    encoder_hidden_states, encoder_hidden_states_mask
)
image_rotary_emb = self.pos_embed(img_shapes, max_txt_seq_len=text_seq_len, device=hidden_states.device)
```

The diffusers **pipeline** does not pass `txt_seq_lens` into the transformer; RoPE length is inferred from the padded tensor geometry.

## Why this harms precision

`QwenEmbedRope` in vllm-omni slices text frequencies with:

```python
max_len = max(txt_seq_lens)
txt_freqs = self.pos_freqs[max_vid_index : max_vid_index + max_len, ...]
```

Joint attention applies those frequencies to **all** `encoder_hidden_states.shape[1]` text positions (`qwen_image_transformer.py`).

If `max(txt_seq_lens) < encoder_hidden_states.shape[1]`:

1. The RoPE table is too short for the padded embedding width.
2. On the native CPU RoPE path this can raise a sequence-length `RuntimeError` during rotary application.
3. When execution proceeds (or on other backends), positions beyond `max(txt_seq_lens)` do not receive the same RoPE as diffusers — attention logits and denoiser outputs diverge from the reference.

**Concrete example:** 200 valid tokens in a tensor padded to width 1058 → vllm-omni uses RoPE length 200, diffusers uses 1058.

For positions `0..199` the frequency *values* match (same offset into `pos_freqs`), but positions `200..1057` are missing RoPE entirely in vllm-omni while diffusers still applies positional encoding across the full padded width (padding tokens are masked in attention but still participate in RoPE indexing consistent with diffusers).

### Non-CB (`forward()` / `diffuse()`)

Stock `encode_prompt()` usually produces tensors where `prompt_embeds.shape[1] == prompt_embeds_mask.sum()` for a single naturally-encoded prompt (no extra right-padding). In that narrow case `mask.sum()` and `shape[1]` agree and the bug is latent.

The mismatch appears whenever embedding width exceeds valid token count, e.g.:

- Caller-supplied `prompt_embeds` padded to a fixed `max_model_len`
- Batched / collated training tensors with fixed sequence width
- Any path that right-pads embeddings without making `mask.sum() == shape[1]`

### Continuous batching (CB)

1. `prepare_encode()` stores `txt_seq_lens = [mask.sum()]`.
2. `InputBatch._prepare_request_prompt_field()` may pad `state.prompt_embeds` / masks to a shared `target_seq_len` **without updating** `state.txt_seq_lens`.
3. `denoise_step()` passes the stale `input_batch.txt_seq_lens` into the transformer.

Even when batching a short request with a longer one, `max(stored txt_seq_lens)` often equals the padded width; the failure mode is when `max(txt_seq_lens)` stays below the padded embed width (e.g. single padded request, fixed-width training pad, or stale per-request values).

### CFG (classifier-free guidance)

Both branches are affected independently — positive and negative `txt_seq_lens` / `negative_txt_seq_lens` are each computed via `mask.sum()` in `_prepare_generation_context()` and passed through:

- non-CB: `cfg_parallel.py::diffuse()` (`positive_kwargs` / `negative_kwargs`)
- CB: `denoise_step()` → `_build_denoise_kwargs()` → `predict_noise_maybe_with_cfg()`

Short negative prompts with padded width exhibit the same RoPE length mismatch as the positive branch.

## Suggested fix

Align with diffusers / the verl-omni stepwise workaround:

```python
# In _prepare_generation_context() (all QwenImage pipeline variants):
txt_seq_lens = [int(prompt_embeds.shape[1])] * int(prompt_embeds.shape[0])
if negative_prompt_embeds is not None:
    negative_txt_seq_lens = [int(negative_prompt_embeds.shape[1])] * int(negative_prompt_embeds.shape[0])
else:
    negative_txt_seq_lens = None
```

Optionally also refresh `state.txt_seq_lens` inside `_prepare_request_prompt_field()` after CB padding so request state stays self-consistent.

Alternative longer-term: stop threading `txt_seq_lens` through the pipeline and infer RoPE length inside `QwenImageTransformer2DModel.forward()` the way diffusers does.

## Impact

- Numerical divergence vs diffusers / PyTorch reference for padded prompts
- RL training: elevated `actor/ppo_kl` at step 1 when comparing vLLM rollout to trainer log-prob recomputation (reported downstream in verl-omni stepwise adapters; fixed there by overriding `txt_seq_lens`, but stock vllm-omni `prepare_encode()` remains affected)

## Environment

- vllm-omni: `main` @ `a693ae67` (2026-06-15)
- diffusers: `main` @ `d1f8e55c3`
- CPU verification: conda env `torch_2.9.0`

### Your code version

<details>
<summary>The commit id or version of vllm</summary>

```text

```
</details>
<details>
<summary>The commit id or version of vllm-omni</summary>

```text

```
</details>

### 🐛 Describe the bug

see above

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by SamitHuang**
## Summary

Fixes #4443.

Qwen-Image pipelines previously derived `txt_seq_lens` from `prompt_embeds_mask.sum(dim=1)`, which counts valid tokens only. Diffusers (and the RL trainer) derive text RoPE length from the **padded encoder hidden-state width** (`prompt_embeds.shape[1]`). When prompts are right-padded for request batching or fixed-width collation, vLLM-Omni built a too-short RoPE table and attention numerics diverged from the reference.

This PR:
- Adds `txt_seq_lens_from_embeds()` and uses it in all Qwen-Image pipeline variants (`pipeline_qwen_image*.py`).
- Refreshes per-request `txt_seq_lens` / `negative_txt_seq_lens` on the continuous-batching path after in-place prompt padding in `input_batch.py`.

## Why this is not duplicating an existing PR

Checked open PRs for `txt_seq_lens` / RoPE / #4443 — no overlapping fix found.

### 1. Previous failure case (inputs)

Representative **request-batch / CB padding** scenario (two requests fused to a shared width):

| Field | Request A | Request B |
|---|---|---|
| Valid tokens (`mask.sum()`) | 10 | 20 |
| Padded width (`prompt_embeds.shape[1]`) | 32 | 32 |

**Pre-PR metadata (buggy):**
```python
txt_seq_lens = prompt_embeds_mask.sum(dim=1).tolist()  # [10, 20]
max(txt_seq_lens) = 20  <  shape[1] = 32  # RoPE table too short
```

**Post-PR metadata (fixed):**
```python
txt_seq_lens = txt_seq_lens_from_embeds(prompt_embeds)  # [32, 32]
max(txt_seq_lens) = 32  ==  shape[1] = 32
```

**RoPE table length (`QwenEmbedRope`, CPU):**

| `txt_seq_lens` | Text freq rows | Matches padded width? |
|---|---|---|
| `[10, 20]` (pre-PR) | **20** | no (20 < 32) |
| `[32, 32]` (post-PR) | **32** | yes |

Single-request CB padding is also covered by `test_input_batch_txt_seq_lens.py` (valid=10, pad to 32).

### 2. Automated tests (CI, CPU)

```bash
python3 -m pytest \
  tests/diffusion/models/qwen_image/test_qwen_image_rope_utils.py \
  tests/diffusion/models/qwen_image/test_qwen_image_rope_table_failure_case.py \
  tests/diffusion/test_input_batch_txt_seq_lens.py -q
```

**Result (2026-06-16): 8 passed**

| Test file | What it verifies |
|---|---|
| `test_qwen_image_rope_utils.py` | Helper uses `shape[1]`, not `mask.sum()` |
| `test_qwen_image_rope_table_failure_case.py` | Documents pre/post `txt_seq_lens` and RoPE table rows for `[10,20]→32` |
| `test_input_batch_txt_seq_lens.py` | CB padding refreshes `state.txt_seq_lens` to padded width |

### 3. Forward numerics vs diffusers (manual, GPU)

One transformer step with **fixed** padded tensors (not run in default CI):

```python
# B=2, T_pad=32, valid=[10, 20], seed=42
# Compare vllm-omni QwenImageTransformer2DModel noise_pred:
#   txt_seq_lens=[10, 20]  (pre-PR path)  -> diverges from diffusers
#   txt_seq_lens=[32, 32]  (post-PR path)  -> matches diffusers
```

Requires local `Qwen/Qwen-Image` weights and initialized vLLM distributed context. Recommended as a reviewer spot-check when GPU is available; not required for CI because the root cause is incorrect metadata and is covered by sections 1–2.

### 4. Generated images (manual, optional)

Existing image accuracy e2e (`tests/e2e/accuracy/test_qwen_image.py`, SSIM/PSNR vs diffusers) **does not trigger this bug** for typical single-prompt runs where `mask.sum() == shape[1]`.

To observe image-level impact you must exercise a **padded embed path**, e.g.:
- request batching with unequal prompt lengths in one admission window, or
- injecting pre-padded `prompt_embeds` via a custom pipeline.

We did **not** attach image artifacts to this PR because the fix is a metadata/RoPE-width correction; image SSIM on natural unpadded prompts is unchanged. Forward `noise_pred` parity (section 3) is the appropriate numerics check for padded batches.

## AI assistance

AI assistance (GitHub Copilot) was used to implement and test this change. A human submitter reviewed the diff.

### CODE REVIEW

File: vllm_omni/diffusion/models/qwen_image/rope_utils.py
Comment (NumberWan) [inline vllm_omni/diffusion/models/qwen_image/rope_utils.py:1]: I am not sure the file may fit better under `vllm_omni/diffusion/utils/`, or directly add `txt_seq_lens_from_embeds()` into prompt_utils.py, since all qwen files already imported prompt_utils, and only the qwen model uses it.

Comment (SamitHuang) [reply vllm_omni/diffusion/models/qwen_image/rope_utils.py:1]: `txt_seq_lens_from_embeds()` is not generic RoPE math. it encodes one Qwen-Image-specific contract

**Review by hsliuustc0106 — CHANGES_REQUESTED** [review]
Platform CI failures (amd-ci and intel-ci). The code fix for RoPE table length mismatch looks correct, but please check the platform CI failures before merging.

**Comment by NumberWan** [comment]
Anchor prompt (accuracy-test default), seed=42, 512×512, 20 steps.  
**diffusers** solo as reference; **vLLM batch** = anchor + 7 concurrent unequal-length fillers.  
Pixel diff vs diffusers (lower = closer):

| vs diffusers | mean_abs_diff | p99 | max  |
|---|---:|---:|---:|
| vLLM batch (**pre-fix**) | 0.0127 | 0.176 | 0.855 |
| vLLM batch (**post-fix**) | 0.0095 | 0.110 | 0.690 |

Pre-fix batch drifts further from diffusers; post-fix moves closer (#4443). CPU tests 8/8 passed locally.

LGTM.

**Comment by SamitHuang** [comment]
> Platform CI failures (amd-ci and intel-ci). The code fix for RoPE table length mismatch looks correct, but please check the platform CI failures before merging.

The amd and intel ci errors are not related to the code changes in this PR. @hsliuustc0106 PTAL

**Review by hsliuustc0106 — APPROVED** [review]

**Review by ybnbarecords-cloud — APPROVED** [review]
